pre-aio-node-with-zuul-ref.sh

```diff
@@ -0,0 +1,157 @@
+#!/bin/bash
+
+set -ex
+refspec="$1"
+ZUUL_CHANGES="$2"
+keep_env="$3"
+
+FETCH_URL=http://10.71.212.50/p/
+ROOT_DIR=/opt/stack/
+DEVSTACK_TOOL=/opt/stack/devstack
+NOVA_DIR=/opt/stack/nova
+TEMPEST_DIR=/opt/stack/tempest
+TMP_MERGE_BRANCH=tmp_branch_for_vgpu_patches
+DEST=${DEST:-"/opt/stack/openstack"}
+NOVA_CONF_PATH=/etc/nova
+SAMPLE_IMAGE=cirros-0.3.5-x86_64-disk
+VM_NAME="testVM"
+# Set up internal variables
+_SSH_OPTIONS="\
+    -o BatchMode=yes \
+    -o StrictHostKeyChecking=no \
+    -o UserKnownHostsFile=/dev/null \
+    -i /opt/stack/.ssh/id_rsa"
+
+server_info=$(grep "VNCSERVER_PROXYCLIENT_ADDRESS" local.conf)
+server_info=${server_info#*\"}
+server_info=${server_info%\"}
+echo "servier ip:"
+echo $server_info
+
+function on_xenserver() {
+    ssh $_SSH_OPTIONS "root@$server_info" bash -s --
+}
+
+cd $DEVSTACK_TOOL/
+./clean.sh
+git checkout master
+git pull --ff origin master
+
+sed -i '/[[local|localrc]]/a DEST=/opt/stack/openstack' ${DEVSTACK_TOOL}/local.conf
+
+IFS='^' read -r -a ZUUL_PROJ_CHANGE <<< "$ZUUL_CHANGES"
+related_proj=()
+for proj in ${ZUUL_PROJ_CHANGE[@]}
+do
+    proj=${proj%%:*}
+    proj=${proj##*/}
+    related_proj+=($proj)
+done
+unique_proj_list=$(echo "${related_proj[@]}" | tr ' ' '\n' | nl | sort -u -k2 | sort -n | cut -f2-)
+
+echo ${DEST}
+cd ${DEST}
+
+for dir in ${DEST}/*
+do
+    if [ -d "${dir}" ] ; then
+        echo "check ${dir}"
+        cd ${dir}
+        if [ -d .git ]; then
+            echo "update repository ${dir}"
+            git checkout master
+            git pull --ff origin master
+
+            pkg_flag=${dir##*/}
+            echo $pkg_flag
+            if [[ " ${unique_proj_list[@]} " =~ " ${pkg_flag} " ]]; then
+                echo "merge repository ${dir}"
+                git branch -D $TMP_MERGE_BRANCH
+                git checkout master -B $TMP_MERGE_BRANCH
+                git fetch ${FETCH_URL}/openstack/${pkg_flag} $refspec
+                git merge FETCH_HEAD -m "temp branch for vgpu patches merge"
+            fi
+        fi;
+    fi
+done
+
+first_vgpu_type=$(on_xenserver << END_OF_REQ_VGPU_TYPE
+vgpu_type_list=\$(xe vgpu-type-list --minimal)
+vgpu_type_list=\${vgpu_type_list//,/ }
+for id in \$vgpu_type_list
+do
+    name=\$(xe vgpu-type-param-get uuid=\$id param-name="model-name")
+    if [ "\$name" != "passthrough" ]
+    then
+        echo \$name
+        exit
+    fi
+done
+END_OF_REQ_VGPU_TYPE
+)
+
+cd $DEVSTACK_TOOL/
+./stack.sh
+echo "[devices]" >> ${NOVA_CONF_PATH}/nova.conf
```

Comment:
suggest put the setting in local.conf before stack.sh, so that there is no need to restart compute service. e.g.
[[post-config|$NOVA_CONF]]
[devices]
enabled_vgpu_types = ....

Reply:
Will move it